pkg/translate/translateValue.go

```diff
@@ -0,0 +1,554 @@
+func (t *ValueYAMLTranslator) TranslateTree(valueTree map[string]interface{}, cpSpecTree map[string]interface{}, path util.Path) error {
+	// translate enablement and namespace
+	err := t.setEnablementAndNamespacesFromValue(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating enablement and namespace from value.yaml tree: %v", err.Error())
+	}
+	// translate with api mapping
+	err = t.translateTree(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating value.yaml tree with global mapping: %v", err.Error())
+	}
+
+	// translate with k8s mapping
+	t.initK8SMapping(valueTree)
+	err = t.translateK8sTree(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating value.yaml tree with kubernetes mapping: %v", err.Error())
+	}
+
+	// translate remaining untranslated paths into component values
+	err = t.translateRemainPaths(valueTree, cpSpecTree, nil)
+	if err != nil {
+		return fmt.Errorf("error when translating remaining path: %v", err.Error())
+	}
+	return nil
+}
+
+// setEnablementAndNamespaces translates the enablement and namespace value of each component in the baseYAML values
+// tree, based on feature/component inheritance relationship.
+func (t *ValueYAMLTranslator) setEnablementAndNamespacesFromValue(valueSpec map[string]interface{}, root map[string]interface{}) error {
+	for cnv, cni := range t.ValuesToComponentName {
+		enabled, err := name.IsComponentEnabledFromValue(cnv, valueSpec)
+		if err != nil {
+			return err
+		}
+		featureName := name.ComponentNameToFeatureName[cni]
+		tmpl := componentEnablementPattern
+		ceVal := renderFeatureComponentPathTemplate(tmpl, featureName, cni)
+		outCP := util.ToYAMLPath(ceVal)
+		// set component enablement
+		if err := tpath.WriteNode(root, outCP, enabled); err != nil {
+			return err
+		}
+		// set feature enablement
+		feVal := featureName + ".Enabled"
+		outFP := util.ToYAMLPath(string(feVal))
+		curEnabled, found, _ := name.GetFromTreePath(root, outFP)
+		if !found {
+			if err := tpath.WriteNode(root, outFP, enabled); err != nil {
+				return err
+			}
+		} else if curEnabled == false && enabled {
+			if err := tpath.WriteNode(root, outFP, enabled); err != nil {
+				return err
+			}
+		}
+	}
+
+	// set namespace
+	for vp, nsList := range t.NamespaceMapping {
+		namespace, err := name.NamespaceFromValue(vp, valueSpec)
+		if err != nil {
+			return err
+		}
+		for _, ns := range nsList {
+			if err := tpath.WriteNode(root, util.ToYAMLPath(ns), namespace); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// translateHPASpec translates HPA related configurations from helm values.yaml tree
+func translateHPASpec(inPath string, outPath string, value interface{}, valueTree map[string]interface{}, cpSpecTree map[string]interface{}) error {
+	asEnabled, ok := value.(bool)
+	if !ok {
+		return fmt.Errorf("expect autoscaleEnabled node type to be bool but got: %T", value)
+	}
+	if asEnabled {
```

Comment:
Keep indentation to a minimum. This can be rewritten as:
if !asEnabled {
  return nil
}
// asEnabled is true here

Reply:
done